### PR TITLE
Write maid scripts to package.json

### LINKED ISSUE #24: Add the ability to sync npm scripts
Introduce an argument `--update-scripts` to write all the maid tasks to the package.json file. 

So `## clean` -> `"scripts: { "clean": "maid clean" }`

This will allow users to use `yarn clean` or `npm run clean` as they may likely already be doing, but having the source of truth be their maid file. I envision that the `--update-scripts` arg would be used with something like lint-staged to automatically update the scripts when there were changes. It would only update the scripts whose commands start with `maid`.

### PR DESCRIPTION
**Pull request by zephraph**
Fixes #24 

Adds the ability to write all maid tasks out to the package.json for easier consumption of tasks. 

- [x] add cli command
- [x] implement script writer
- [x] document command
- [x] add lint-staged example
- [x] dogfood setup
- [x] account for pre/post scripts
- [ ] tests

### CODE REVIEW

**Comment by zephraph** [comment]
Okay, so I've made definite progress. The core of the functionality is there and the project itself is using it. There are still some outstanding issues though. 

1. when pairing this up with `lint-staged` the maidfile is what will trigger the script update... but the package.json is the file that will change. How do we signify that a `git add` should happen?
2. I implemented this as a global option, but I'm not convinced it shouldn't just be a command... I started down that path, but I backtracked because I didn't want to pollute the command space.

File: README.md [hunk 1]
```diff
@@ -281,6 +282,8 @@ If you want to automatically fix lint errors, try adding `--fix` plugin to the c
 
 Use [AVA](https://github.com/avajs/ava) to run unit tests.
 
+Run task `lint` before this.
```
Comment (zephraph) [inline README.md]: This is the only real change in this file... the rest is prettier.

File: bin/cli.js [hunk 2]
```diff
@@ -1,8 +1,18 @@
 #!/usr/bin/env node
 const cli = require('cac')()
 const chalk = require('chalk')
+const MaidError = require('../lib/MaidError')
 
 cli.command('*', 'Run a task in current working directory', (input, flags) => {
+  if (flags.updateScripts && input[0]) {
+    throw new MaidError('Cannot run task and update scripts')
```
Comment (egoist) [inline bin/cli.js]: why disallow this 🤔

Comment (zephraph) [reply bin/cli.js]: in the updateScripts method, I'm passing all the arguments through transparently. If I allowed this I'd have to find the input and remove it before writing to the package.json. 

So... laziness. I can remove it. Point might be moot if this turns into its own command though.

File: package.json [hunk 3]
```diff
@@ -63,7 +65,8 @@
     ],
     "README.md": [
       "node bin/cli toc",
+      "node bin/cli update-scripts --git-add -p",
```
Comment (zephraph) [inline package.json]: magic

**Comment by zephraph** [comment]
Thought of another issue. `pre` and `post` tasks will conflict with the package.json's version. 

Let's say there are two tasks... `prebuild` and `build`. Running `yarn build` would trigger `yarn prebuild` which would run the `prebuild` task... but that task would be ran again by maid before `build` started. So we may need to filter those out before writing the tasks.

**Comment by olstenlarck** [comment]
Remove the implementation of `pre` and `post` and rely on the scripts for that?

Few positives from that:

1. lower code
2. battle tested handling
3. better "out of the box behaving" - so no need to write "run that before this and after this" kinda sentences. Just say users that if they want hooks, they can use: a] package scripts for hooks - so will be able to run `maid build` and that will executethe `prebuild` script that points to `maid test`, or b] just call the desired tasks in what order they want - e.g. `maid test build` to run `test` then `build` task.

The `3.a` may be a bit trickier probably, but second one is whole lot easier and obvious and familiar.

So let's rephrase...

### Scenario 1

User syncs scripts. 

```json
{
  "scripts": {
    "test": "maid test",
    "build": "maid build"
  }
}
```

If he want to run `test` before `build` would just run `maid test build` (`npx maid test build` or `yarn maid test build`) which requires the implementation for running multiple tasks, a la gulp? Btw, comma separated tasks.

It's clean, obvious, familiar, less verbose (on tasks and scripts definition) and the multiple tasks impl would be smaller.

### Scenario 2

User syncs the scripts (with that PR's command) with specific flag (can't come name for that flag to my mind ;d). This command is implemented with the current idea of defining sentences, respect them (only?) when ran with that flag... 

```
maid update-scripts --hooks
```

generates 

```json
{
  "scripts": {
    "prebuild": "maid test",
    "build": "maid build"
  }
}
```

from this maid file

```markdown
## build

Build our main app

<!-- Following line is a maid command for running task -->
Run task `test` before this

\`\`\`bash
# note that you can directly call binaries inside node_modules/.bin
# just like how `npm scripts` works
babel src -d lib
\`\`\`

## test

Test all the things.

\`\`\`bash
ava tests/**/*.js
\`\`\`

foo bar

```


doh how ugly is my last code block... how you handle the markdown in markdown situation, here in github issues?! :rofl:

**Comment by zephraph** [comment]
@egoist I would actually be all for removing the `pre` and `post` commands... but I understand why they're there and I'm not looking to make that change in this PR. 

I added logic to check every task that starts with `pre` or `post` against the list of tasks. If the a task exists that is the representative base task (i.e minus `pre` or `post`) then it won't write the task. 

That means if there's a `prebuild` and `build` command then only `build` will be written. But if a `precommit` task exists, but `commit` doesn't, then `precommit` will still be written.

**Comment by zephraph** [comment]
@egoist I still need to add tests but this works/is ready for review.

**Pull request closed without merging**